src/pcp.rs

```diff
@@ -135,65 +199,269 @@ pub trait ValueParam: Clone + Debug {
-    /// shares (`num_shares`).
-    fn new_share(
-        data: Vec<Self::Field>,
-        param: &Self::Param,
-        num_shares: usize,
-    ) -> Result<Self, PcpError>;
+                Ok(gadget)
+            })
+            .collect::<Result<Vec<_>, PcpError>>()?;
+
+        // Create a buffer for storing the proof. The buffer is longer than the proof itself; the extra
+        // length is to accommodate the computation of each gadget polynomial.
+        let data_len = (0..shim.len())
+            .map(|idx| {
+                shim[idx].arity() + shim[idx].degree() * (1 + shim[idx].calls()).next_power_of_two()
+            })
+            .sum();
+        let mut proof = vec![Self::Field::zero(); data_len];
+
+        // Run the validity circuit with a sequence of "shim" gadgets that record the value of each
+        // input wire of each gadget evaluation. These values are used to construct the wire
+        // polynomials for each gadget in the next step.
+        let _ = self.valid(&mut shim, input, joint_rand, 1);
+
+        // Fill the buffer with the proof. `proof_len` keeps track of the amount of data written to the
+        // buffer so far.
+        let mut proof_len = 0;
+        for idx in 0..shim.len() {
+            let gadget = shim[idx]
+                .as_any()
+                .downcast_mut::<ProveShimGadget<Self::Field>>()
+                .unwrap();
+
+            // Interpolate the wire polynomials `f[0], ..., f[g_arity-1]` from the input wires of each
+            // evaluation of the gadget.
+            let m = (1 + gadget.calls()).next_power_of_two();
+            let m_inv =
+                Self::Field::from(<Self::Field as FieldElement>::Integer::try_from(m).unwrap())
+                    .inv();
+            let mut f = vec![vec![Self::Field::zero(); m]; gadget.arity()];
+            for wire in 0..gadget.arity() {
+                discrete_fourier_transform(&mut f[wire], &gadget.f_vals[wire], m)?;
+                discrete_fourier_transform_inv_finish(&mut f[wire], m, m_inv);
+
+                // The first point on each wire polynomial is a random value chosen by the prover. This
+                // point is stored in the proof so that the verifier can reconstruct the wire
+                // polynomials.
+                proof[proof_len + wire] = gadget.f_vals[wire][0];
+            }
+
+            // Construct the gadget polynomial `G(f[0], ..., f[g_arity-1])` and append it to `proof`.
+            gadget.call_poly(&mut proof[proof_len + gadget.arity()..], &f)?;
+            proof_len += gadget.arity() + gadget.degree() * (m - 1) + 1;
+        }
 
-    /// Evaluates the validity circuit on the given input (i.e., `self`) and returns the output.
-    /// `joint_rand` is the joint randomness shared by the prover and verifier. `g` is the sequence
-    /// of gadgets called by the circuit.
+        // Truncate the buffer to the size of the proof.
+        proof.truncate(proof_len);
```

Comment:
I'm kind of surprised we never use `self.proof_len()` in this function. Should we at least check that the `proof_len` computed here is the expected value?

Reply:
I agree with your observation that this is programmer error so a panic feels appropriate.

Reply:
Done.